tidb/src/tidb/util.clj

```diff
@@ -35,3 +36,44 @@
        (map #(str/split % #"=" 2))
        (filter #(not (empty? (second %))))
        (into {})))
+
+(defn get-start-ts [op] (get-in op [:txn-info :start_ts] 0))
+(defn get-commit-ts [op] (get-in op [:txn-info :commit_ts] 0))
+
+(defrecord TSOExplainer []
+  DataExplainer
+  (explain-pair-data
+    [_ a b]
+    {:type :tso :a a :b b})
+  (render-explanation
+    [_ {:keys [a b]} a-name b-name]
+    (str a-name "'s commit-ts " (get-commit-ts a) " < " b-name "'s start-ts " (get-start-ts b))))
+
+(defn tso-graph [history]
```

Comment:
Do we need to add a test case for it?

Reply:
Do you mean unit test? I just test it manually. The whole [tidb sub-project](https://github.com/pingcap/jepsen/tree/testing_ci/tidb) doesn't have unit test and there is no CI for it.

The graph will be used in existing append and monotonic workloads (also introduced in txn-cycle, but this workload is not scheduled yet).

Reply:
We could also add some explanations in the  [document](https://docs.google.com/document/d/1v69-PiN0XGV64xwuRVWHAPvq3AWY7BdC-1WEISidh74/edit) about it.